### PR TITLE
2267 add services to machine

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do?

Fixes services part of #2267

Add any further explanations here.

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running unit tests
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py [hunk 1]
```diff
@@ -88,6 +94,17 @@ def _update_target_machine_os(self, machine: Machine, event: PingScanEvent):
...
+    def _update_network_services(self, target: Machine, event: TCPScanEvent):
+        network_services = {
+            SocketAddress(ip=event.target, port=port): NetworkService.UNKNOWN
+            for port in self._get_open_ports(event)
+        }
+        self._machine_repository.upsert_network_services(target.id, network_services)
+
+    @staticmethod
+    def _get_open_ports(event: TCPScanEvent) -> List[int]:
+        return [port for port, status in event.ports.items() if status == PortStatus.OPEN]
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py:106]: This should return list of `NetworkPort`.

File: monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py [hunk 2]
```diff
@@ -110,7 +109,7 @@ def _update_nodes(self, target_machine: Machine, event: ScanEvent):
 
     def _update_tcp_connections(self, src_node: Node, target_machine: Machine, event: TCPScanEvent):
         tcp_connections = set()
-        open_ports = (port for port, status in event.ports.items() if status == PortStatus.OPEN)
+        open_ports = self._get_open_ports(event)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py:117]: We can store the list of ports on line 52 and pass that list to this function instead of calling `_get_open_ports()` again.

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py [hunk 3]
```diff
@@ -38,12 +38,6 @@ def upsert_machine(self, machine: Machine):
         except Exception as err:
             raise StorageError(f'Error updating machine with ID "{machine.id}": {err}')
 
-        if result.matched_count != 0 and result.modified_count != 1:
-            raise StorageError(
-                f'Error updating machine with ID "{machine.id}": Expected to update 1 machine, '
-                f"but {result.modified_count} were updated"
-            )
-
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: > It doesn't make sense to raise an error if upsert did no changes, because the purpose of "upsert" method is to ensure that data is up-to-date. If no changes were made it means it's already up-to-date.

Should upsert have even been called then? 
Let's at least add a log statement regarding this. This seems like useful debugging information if we run into bugs.

File: monkey/monkey_island/cc/repository/utils.py [hunk 4]
```diff
@@ -33,3 +35,34 @@ def initialize_machine_repository(machine_repository: IMachineRepository):
...
+    replaces all occurances of "." with ",,,"
+    :param mapping: Mapping to be converted to mongo compatible mapping
+    :return: Mongo compatible mapping
+    """
+    mapping_json = json.dumps(mapping)
+    if DOT_REPLACEMENT in mapping_json:
+        raise StorageError(
+            f"Mapping {mapping} already contains {DOT_REPLACEMENT}."
+            f" Aborting the encoding procedure"
+        )
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/repository/utils.py:55]: Is this necessary? The next line will just move on silently if it finds no dots.

Comment (shreyamalviya) [reply monkey/monkey_island/cc/repository/utils.py:55]: If I'm missing something and it is necessary, update the docstring to mention that this error could be raised.

File: monkey/monkey_island/cc/models/report/report_dal.py [hunk 5]
```diff
@@ -23,6 +23,7 @@ def get_report() -> dict:
 
 # TODO remove this unnecessary encoding. I think these are legacy methods from back in the day
 # when usernames were used as keys. If not, we shouldn't use unknown data as keys.
+# Already implemented in monkey_island/cc/repository/utils.py
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/models/report/report_dal.py:26]: And we're not removing it because?

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/models/report/report_dal.py (hunk 5 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/models/report/report_dal.py:26]: We're not done with the report yet

File: monkey/monkey_island/cc/repository/utils.py (hunk 4 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/utils.py:55]: Raising an error is necessary, because we can't encode and decode a string like "these are three commas: ,,,". Originally the string was "these are three commas: ,,,", but on decode it will be "these are three commas: .". Whatever we use as a replacement for `.` can't be in the original string or we lose data

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py (hunk 3 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: Yes, it should've been called then, because the alternative is double the amount of queries (you check if DB is up to date and then update it) for the same result(DB is up to date).

Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: We would need to log everything then. Upserting when the machine is already up to date seems like the most common and least valuable information. So far I think whatever receives the error should be logging it, because that's the information we care about

File: monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py (hunk 2 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py:117]: Information about open ports is already in the event. The way to go is to probably build a handler with unpacked values like target, open ports, etc. methods shouldn't have 10 parameters that are duplicate data.
Anyways, Mike said he'll have to refactor this anyway, so I'm not worried about multiple calls to the same method

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py (hunk 3 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: Fair

**Review by shreyamalviya — APPROVED** [review]

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py [hunk 6]
```diff
@@ -32,26 +34,32 @@ def get_new_id(self) -> MachineID:
...
+            self.upsert_machine(machine)
+        except Exception as err:
+            raise StorageError("Failed upserting the machine or adding services") from err
+
     def get_machine_by_id(self, machine_id: MachineID) -> Machine:
-        return self._find_one("id", machine_id)
+        machine = self._find_one("id", machine_id)
+        if not machine:
+            raise UnknownRecordError(f"Machine with id {machine_id} not found")
+        return machine
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/mongo_machine_repository.py:62]: Why this change? `_find_one()` does not return `None`.

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py [hunk 7]
```diff
@@ -75,18 +84,18 @@ def get_machines(self) -> Sequence[Machine]:
         except Exception as err:
             raise RetrievalError(f"Error retrieving machines: {err}")
 
-        return [Machine(**m) for m in cursor]
+        return [Machine(**mongo_dot_decoder(m)) for m in cursor]
 
     def get_machines_by_ip(self, ip: IPv4Address) -> Sequence[Machine]:
-        ip_regex = "^" + str(ip).replace(".", "\\.") + "\\/.*$"
+        ip_regex = "^" + str(ip).replace(".", DOT_REPLACEMENT) + "\\/.*$"
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/mongo_machine_repository.py:90]: Can we do

```suggestion
        ip_regex = "^" + mongo_dot_encoder(str(ip)) + "\\/.*$"
```

File: monkey/monkey_island/cc/repository/utils.py [hunk 8]
```diff
@@ -33,3 +35,34 @@ def initialize_machine_repository(machine_repository: IMachineRepository):
             hostname=gethostname(),
         )
         machine_repository.upsert_machine(machine)
+
+
+DOT_REPLACEMENT = ",,,"
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/utils.py:40]: We should consider something more unique than 3 commas. If the data is in CSV format, it's very likely there will be 3 commas in a row (empty fields) and the data won't encode.

We can use a string like "[MONGO_DOT_REPLACEMENT]". It's less likely you'll find that string in any valid data. We could add other stuff to make it even less likely. "[INFECTION_MONKEY_MONGO_DOT_REPLACEMENT]".

You get the idea.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py (hunk 3 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: I'm also not sure when would this case apear? when we accidentaly upsert more machines than aticipated?

Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: It would happen if there's some weird mongo error that didn't raise an exception :shrug:. I can't tell from the documentation if it's possible, so it's safer for us to handle it.

Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: This is what I get though:
image

Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py:46]: That's the only thing I've found in the docs: 
image

Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py]: That seems to indicate that my assumption based on the documentation was wrong.

It looks like the best thing to do drop https://github.com/guardicore/monkey/pull/2409/commits/d71791750a038d3793cb6fe12df6478802649c82

Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py]: On `develop`, I added a unit test for the same condition in `MongoAgentRepository` and removed this check. See the commit message on https://github.com/guardicore/monkey/commit/7fc0aaf42b80d592430d8bc9263ee61777095dca and the diff for more details.

Comment (shreyamalviya) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py]: > though I'm not sure why I was getting the 0's
See https://www.mongodb.com/docs/manual/reference/method/db.collection.replaceOne/#replace-with-upsert. Maybe there's something different in what you're upserting, ie, it's a new document.

File: monkey/monkey_island/cc/repository/utils.py (hunk 8 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/utils.py:40]: I wanted to leave it consistent with `_encode_dot_char_before_mongo_insert`. We can make it more unique, but it should be short in order to for keys to still be readable (if there's an IP address where each dot is replaced with "[INFECTION_MONKEY_MONGO_DOT_REPLACEMENT]" you'd need to do quire a bit of scrolling). I'd suggest something like `[_D0T_]`

File: monkey/monkey_island/cc/repository/mongo_machine_repository.py (hunk 3 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py]: [This](https://stackoverflow.com/a/39714522) may have something to do with it.

Comment (shreyamalviya) [reply monkey/monkey_island/cc/repository/mongo_machine_repository.py]: [This](https://stackoverflow.com/a/39714522) may be useful.

File: monkey/monkey_island/cc/repository/utils.py (hunk 8 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/utils.py]: `[D0T]` sounds good to me.
